File: BackTracking/NQueens/NQueens.py
from typing import List

class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def isSafe(row: int, col: int, board: List[str]) -> bool:
            # Check the left side in the same row
            for i in range(col):
                if board[row][i] == 'Q':
                    return False

            # Check upper-left diagonal
            i, j = row, col
            while i >= 0 and j >= 0:
                if board[i][j] == 'Q':
                    return False
                i -= 1
                j -= 1

            # Check lower-left diagonal
            i, j = row, col
            while i < n and j >= 0:
                if board[i][j] == 'Q':
                    return False
                i += 1
                j -= 1

            return True

        def solve(col: int, board: List[str], ans: List[List[str]]):
            # Base case: If all queens are placed
            if col == n:
                ans.append(board.copy())  # Add a valid board configuration to the answer
                return

            # Try placing a queen in each row of the current column
            for row in range(n):
                if isSafe(row, col, board):
                    board[row] = board[row][:col] + 'Q' + board[row][col+1:]  # Place the queen
                    solve(col + 1, board, ans)  # Recur to place the next queen
                    board[row] = board[row][:col] + '.' + board[row][col+1:]  # Backtrack: Remove the queen

        ans = []
        board = ['.' * n for _ in range(n)]  # Create an n x n board filled with '.'
        solve(0, board, ans)  # Start solving from the first column
        return ans

# Main function to test the solution
if __name__ == "__main__":
    n = int(input("Enter the value of n: "))
    solution = Solution()
    solutions = solution.solveNQueens(n)

    # Print all the solutions
    for sol in solutions:
        for row in sol:
            print(row)
        print()  # Print a new line between different solutions

File: BackTracking/NQueens/NQueens.cpp
#include <iostream>
#include <vector>
#include <string> // Include string for string manipulations

using namespace std;

// Function to check if placing a queen at board[row][col] is safe
bool isSafe(int row, int col, vector<string> &board, int n) {
    // Check the left side in the same row
    for (int i = col; i >= 0; i--) {
        if (board[row][i] == 'Q') {
            return false;
        }
    }

    // Check upper-left diagonal
    int i = row, j = col;
    while (i >= 0 && j >= 0) {
        if (board[i][j] == 'Q') {
            return false;
        }
        i--; j--;
    }

    // Check lower-left diagonal
    int p = row, q = col;
    while (p < n && q >= 0) {
        if (board[p][q] == 'Q') {
            return false;
        }
        p++; q--;
    }

    return true;
}

// Backtracking function to solve the N-Queens problem
void solve(int col, vector<string> &board, vector<vector<string>> &ans, int n) {
    // Base case: If all queens are placed
    if (col == n) {
        ans.push_back(board); // Add a valid board configuration to the answer
        return;
    }

    // Try placing a queen in each row of the current column
    for (int row = 0; row < n; row++) {
        if (isSafe(row, col, board, n)) {
            board[row][col] = 'Q'; // Place the queen
            solve(col + 1, board, ans, n); // Recur to place the next queen
            board[row][col] = '.'; // Backtrack: Remove the queen
        }
    }
}

// Main function to initialize the board and start solving
vector<vector<string>> solveNQueens(int n) {
    vector<vector<string>> ans; // To store all the valid solutions
    vector<string> board(n, string(n, '.')); // Create an n x n board filled with '.'

    solve(0, board, ans, n); // Start solving from the first column
    return ans; // Return the list of solutions
}

int main() {
    int n;
    cout << "Enter the value of n: ";
    cin >> n; // Input for the size of the board

    vector<vector<string>> solutions = solveNQueens(n); // Solve the N-Queens problem

    // Print all the solutions
    for (const auto &solution : solutions) {
        for (const auto &row : solution) {
            cout << row << endl; // Print each row of the solution
        }
        cout << endl; // Print a new line between different solutions
    }

    return 0; // Return success
}
